README
=== README for IO-Mux-HTTP version 0.10
=   Generated on Wed Jan 26 16:35:02 2011 by OODoc 1.07

There are various ways to install this module:

 (1) if you have a command-line, you can do:
       perl -MCPAN -e 'install <any package from this distribution>'

 (2) if you use Windows, have a look at http://ppm.activestate.com/

 (3) if you have downloaded this module manually (as root/administrator)
       gzip -d IO-Mux-HTTP-0.10.tar.gz
       tar -xf IO-Mux-HTTP-0.10.tar
       cd IO-Mux-HTTP-0.10
       perl Makefile.PL
       make          # optional
       make test     # optional
       make install

For usage, see the included manual-pages or
    http://search.cpan.org/dist/IO-Mux-HTTP-0.10/

Please report problems to
    http://rt.cpan.org/Dist/Display.html?Queue=IO-Mux-HTTP



ChangeLog

==== version history for IO::Mux::HTTP

version 0.10: Wed Jan 26 16:34:55 CET 2011
	- initial implementation, derived from HTTP::Deamon::Multiplex
	  and some LWP code fragments.


Makefile.PL
#!/usr/bin/env perl
use ExtUtils::MakeMaker;

use 5.008;

my $version = '0.10';

WriteMakefile
 ( NAME	 	=> 'IO::Mux::HTTP'
 , VERSION	=> $version
 , PREREQ_PM    =>
    { Test::More  => 0.80
    , IO::Mux     => 0.11
    , Log::Report => 0.91
    , LWP         => 5.834
    , Time::HiRes => 1.9719
    }
 , EXE_FILES    => []
 , AUTHOR       => 'Mark Overmeer'
 , ABSTRACT     => 'IO::Mux based HTTP'
 , LICENSE      => 'perl'
 );

### used by oodist during production of distribution
sub MY::postamble { <<'__POSTAMBLE' }

# for DIST
RAWDIR          = ../public_html/io-mux-http/raw
DISTDIR         = ../public_html/io-mux-http/source
LICENSE         = artistic

# for POD
FIRST_YEAR      = 2011
EMAIL           = perl@overmeer.net
WEBSITE         = http://perl.overmeer.net/

EXTENDS		= ../AnyDaemon:../IOMux
PODTAIL		= ../daemon-shared-podtail
__POSTAMBLE


examples/client.pl
#!/usr/bin/env perl
# This script demonstrates an HTTP client, to be used with the
# daemon.pl in this same directory.
#
# You may run the test with
#   ./server.pl                # to start the server
#   ./client.pl                # this script
#   kill $(cat server.pl.pid)  # stop server
# Don't forget to look in /var/log/messages!
#
# The server will also work when the client is not multiplexed. The
# multiplexing client will also works when the server is not multiplexed.

use warnings;
use strict;

use Log::Report;

# Any IO::Mux multiplexer implementation (choice independent from
# the choice in the server.
use IO::Mux::Select;
#use IO::Mux::Poll;

use IO::Mux::HTTP::Client;

use Getopt::Long   qw/GetOptions :config no_ignore_case bundling/;
use File::Basename qw/basename/;
use HTTP::Status   qw/HTTP_OK/;

sub start_requesting($);

#
## get command-line options
#

my $mode     = 0;     # increase output

my %net_opts =
  ( host       => 'localhost:5422'
  , port       => undef
  );

GetOptions
   'host|h=s'       => \$net_opts{host}
 , 'port|p=s'       => \$net_opts{port}
 , 'v+'             => \$mode  # -v -vv -vvv
    or exit 1;

unless(defined $net_opts{port})
{   my $port = $net_opts{port} = $1
        if $net_opts{host} =~ s/\:([0-9]+)$//;
    defined $port or error "no port specified";
}

# From now on, all errors and warnings are also sent to syslog,
# provided by Log::Report. Output still also to the screen.
dispatcher SYSLOG => 'syslog', accept => 'INFO-'
  , identity => 'http-c', facility => 'local0';

dispatcher mode => $mode, 'ALL' if $mode;

my $addr   = "$net_opts{host}:$net_opts{port}";
my $client = IO::Mux::HTTP::Client->new
  ( # Configures the server socket
    PeerAddr => $addr
  , Proto     => 'tcp'
  );

# Start any multiplexer.
my $mux    = IO::Mux::Select->new;
#my $mux    = IO::Mux::Poll->new;
$mux->add($client);

start_requesting($client);

$mux->loop;

# All handled
exit 0;

########### Webclient logic

use Data::Dumper;
sub start_requesting($)
{   my $client = shift;
    my $session = {aap => 1};
    my $req1 = HTTP::Request->new(GET => '/any/uri/step1');
    $client->sendRequest($req1, \&step2, $session);
}

sub step2
{   my ($client, $req1, $status, $resp1, $session) = @_;
    print "cl status1 = $status; uri =".$req1->uri."\n";
    for my $c (2..9)
    {   my $reqc = HTTP::Request->new(GET => "req-step$c");
        $client->sendRequest($reqc, \&step3, $session);
    }
}

sub step3
{   my ($client, $reqc, $status, $resp, $session) = @_;
    print "cl status = $status, uri=".$reqc->uri." resp=".$resp->content."\n";
    ++$session->{received} ==8 or return;

    print "all received";
    $client->close;
}

1;


examples/daemon.pl
#!/usr/bin/env perl
# This script demonstrates an HTTP server.
#
# You may run the test with
#   wget http://localhost:5422/any/url
# or with ./client.pl in this same directory.

use warnings;
use strict;

use Log::Report;
use Any::Daemon;

#use IO::Mux::Select;
use IO::Mux::Poll;
use IO::Mux::HTTP::Server;

use Getopt::Long   qw/GetOptions :config no_ignore_case bundling/;
use File::Basename qw/basename/;
use HTTP::Status   qw/HTTP_OK/;

#
## get command-line options
#

my $mode     = 0;     # increase output

my %os_opts  =
  ( pid_file   => basename($0). '.pid'  # usually in /var/run
  , user       => undef
  , group      => undef
  );

my %run_opts =
  ( background => 1
  , max_childs => 1    # there can only be one multiplexer
  );

my %net_opts =
  ( host       => 'localhost:5422'
  , port       => undef
  );

GetOptions
   'background|bg!' => \$run_opts{background}
 , 'group|g=s'      => \$os_opts{group}
 , 'host|h=s'       => \$net_opts{host}
 , 'pid-file|p=s'   => \$os_opts{pid_file}
 , 'port|p=s'       => \$net_opts{port}
 , 'user|u=s'       => \$os_opts{user}
 , 'v+'             => \$mode  # -v -vv -vvv
    or exit 1;

$run_opts{background} //= 1;

unless(defined $net_opts{port})
{   my $port = $net_opts{port} = $1
        if $net_opts{host} =~ s/\:([0-9]+)$//;
    defined $port or error "no port specified";
}

#
## initialize the daemon activities
#

# From now on, all errors and warnings are also sent to syslog,
# provided by Log::Report. Output still also to the screen.
dispatcher SYSLOG => 'syslog', accept => 'INFO-'
  , identity => 'http-s', facility => 'local0';

dispatcher mode => $mode, 'ALL' if $mode;

my $daemon = Any::Daemon->new(%os_opts);

$daemon->run
  ( child_task => \&run_multiplexer
  , %run_opts
  );

exit 1;   # will never be called

sub run_multiplexer()
{
#   my $mux    = IO::Mux::Select->new;
    my $mux    = IO::Mux::Poll->new;

eval {
    # Create one or more listening TCP or UDP sockets.
    my $addr   = "$net_opts{host}:$net_opts{port}";
    my $server = IO::Mux::HTTP::Server->new
      ( name      => 'http-server'
      , handler   => \&incoming_request

        # Configures the server socket
      , LocalAddr => $addr
      , Listen    => 5
      , Proto     => 'tcp'

      );
   $mux->add($server);

   $mux->loop;
};
info "EVAL: $@" if $@;
   exit 0;
}

########### Webserver logic

sub incoming_request($$)
{   my ($client, $req) = @_;
    # trace "INCOMING=".$req->as_string;
    my $resp = $client->makeResponse
      ( $req, HTTP_OK
      , [ Content_Type => 'text/plain' ]
      , "URI=".$req->uri
      );
    $client->sendResponse($resp, \&step2);
}


sub step2
{   my ($client, $resp1, $status, $req2) = @_;
info "se step2 $status ".$resp1->content." ".$req2->uri;

    incoming_request $client, $req2  # not for me: new thread
        if $status==HTTP_OK;
}

1;


lib/IO/Mux/HTTP.pm
# Copyrights 2011 by Mark Overmeer.
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 1.07.
use warnings;
use strict;

package IO::Mux::HTTP;
use vars '$VERSION';
$VERSION = '0.10';

use base 'IO::Mux::Net::TCP';

use Log::Report      'io-mux-http';
use Time::HiRes      qw(time);

use HTTP::Request    ();
use HTTP::Response   ();
use HTTP::Status;
use HTTP::Date       qw(time2str);

use constant
  { HTTP_0_9 => 'HTTP/0.9'
  , HTTP_1_0 => 'HTTP/1.0'
  , HTTP_1_1 => 'HTTP/1.1'
  };

# oops, dirty hack
sub HTTP::Message::id() { shift->{IMH_id} }


my $conn_id = 'C0000000';

sub init($)
{   my ($self, $args) = @_;
    $args->{name} ||= ++$conn_id;

    $self->SUPER::init($args);
    $self->{IMH_headers}   = $args->{add_headers} || [];

    $self->{IMH_requests}  = [];
    $self->{IMH_starttime} = time;
    $self->{IMH_msgcount}  = 0;  # something unique for logging
    $self;
}


sub startTime() {shift->{IMH_starttime}}

sub mux_input($)
{   my ($self, $refdata) = @_;

    while($$refdata)   # possibly more than one message in one TCP package
    {
        # Read header
        my $msg = $self->{IMH_incoming};
        unless($msg)
        {   if($self->{IMH_no_more})
            {   # ignore input for closing, connection can still be writing
                $$refdata = '';
                return;
            }

            $$refdata =~ s/^\s+//s;      # strip leading blanks, sloppy remote
            $$refdata =~ s/(.*?)\r?\n\r?\n//s
                or return;               # not whole header yet, wait for more

            $msg      = $self->{IMH_incoming} = $self->headerArrived($1);
            #trace "new header ".$msg->uri if $msg->isa('HTTP::Request');

            my $msgid = sprintf 'in-%s-%02d'
                , $self->name, $self->{IMH_msgcount}++;
            $msg->id($msgid);
        }

        my $headers = $msg->headers;
        my $proto   = $msg->protocol;

        $msg->protocol($proto = HTTP_0_9)
            unless $proto;

        $self->{IMH_no_more}++
            if $msg->protocol lt HTTP_1_1
            || lc($headers->header('Connection') || '') ne 'keep-alive';

        $self->{IMH_take_all}++
            if $proto lt HTTP_1_0;

        return   # simply wait until EOF
            if $self->{IMH_take_all};

        # Read body

        my $result = $self->bodyComponentArrived($msg, $refdata)
            or return;   # message not ready yet

        my $resp   = $result->isa('HTTP::Response') ? $result : undef;

        $self->shutdown(0)
            if $self->{IMH_no_more};

        delete $self->{IMH_incoming};
        $self->messageArrived($msg, $resp);
    }
}

sub mux_outputbuffer_empty()
{   my $more = shift->{IMH_more_output} or return;
    $more->();
}

sub mux_eof($)
{   my ($self, $refdata) = @_;

    my $msg = delete $self->{IMH_incoming};  # headers only

    if($msg && length($$refdata) && $self->{IMH_take_all})
    {   $msg->content_ref($refdata);
    }
    else
    {   trace "trailing ".length($$refdata)." bytes ignored"
            if $$refdata =~ m/\S/;
    }

    $self->messageArrived($msg)
        if $msg;

    $self->SUPER::mux_eof($refdata);
}

sub bodyComponentArrived($$)
{   my ($self, $msg, $refdata) = @_;

    my $headers = $msg->headers;
    if(my $cl = $headers->header('Content-Length'))
    {   return if length($$refdata) < $cl;   # wait for more
        $msg->content(substr $$refdata, 0, $cl, '');
        return $msg;
    }

    # No Content-Length for multiparts?
    my $ct = $headers->header('Content-Type') || '';
    if($ct =~ m/^multipart\/\w+\s*;.*boundary\s*=(["']?)\s*(\w+)\1/i)
    {   $$refdata =~ s/(.*?\r?\n--\Q$2\E--\r?\n)//
            or return;  # multipart terminator not received yet
        $msg->content($1);
        return $msg;
    }

    # No Content-Length and not multipart, then no body.
    $msg;
}

sub headerArrived($)  {panic}
sub messageArrived($) {panic}

#--------------

sub sendMessage($$)
{   my ($self, $msg, $callback) = @_;

    if($self->mux_output_waiting || $self->{IMH_more_output})
    {   # Arggg. Well, some message content still being written.
        # Do not flood the outbufs with stringified requests.
        # For instance, a number of large files to be sent back
        # or uploaded as chunked.
        push @{$self->{IMH_queued}}, [$msg, $callback];
        return;
    }

    # Write the message now, and after that, but do not forget to
    # handle messages which arrived during this sending after it.
    my $queue_cb;
    $queue_cb = sub
      { my $queued = shift @{$self->{IMH_queued}} or return;
        my ($next_msg, $user_cb) = @$queued;       # the next msg
        $self->writeMessage($next_msg, sub {$queue_cb->(); $user_cb->()});
      };

    $self->writeMessage($msg, $queue_cb);
}

sub writeMessage($$)
{   my ($self, $msg, $callback) = @_;

    my $header = $msg->headers;
    $header->push_header
      ( Date       => time2str(time)
      , Connection => ($self->{IMH_no_more} ? 'close' : 'keep-alive')
      , @{$self->{IMH_headers}}
      );

    my $content = $msg->content;
    if(ref $content eq 'CODE')
    {   # create chunked
        $header->push_header(Transfer_Encoding => 'chunked');
        my $size = 0;
        $self->{IMH_more_output} = sub
          { my $chunk = $content->();
            unless(defined $chunk)
            {  delete $self->{IMH_more_output};
               $self->write("0\r\n\r\n");  # end chunks and no footer
               $size += 5;
               info "sent CHUNKED msg ".$msg->id.' '.$msg->status." ${size}b";
               return $callback->();
            }
            length $chunk or return;
            my $hexlen = sprintf "%x", length $chunk;
            $size     += length($hexlen) + length($chunk) + 4;
            $self->write("$hexlen\r\n$chunk\r\n");
          };
        $self->write(\$header->as_string);
    }
    else
    {   # write message in one go.
        $header->push_header(Content_Length => length $content);
        $msg->content_ref(\$content);
        my $text = $msg->as_string;
        $self->write(\$text);
        info "sent msg ".length($text)."b "
.(ref $msg).' '.($msg->isa('HTTP::Request') ? $msg->uri : $msg->content);
        $callback->();
    }
}


sub closeConnection() { shift->{IMH_no_more} = 1 }

1;


lib/IO/Mux/HTTP/Client.pm
# Copyrights 2011 by Mark Overmeer.
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 1.07.
use warnings;
use strict;

package IO::Mux::HTTP::Client;
use vars '$VERSION';
$VERSION = '0.10';

use base 'IO::Mux::HTTP';

use Log::Report 'io-mux-http';

use HTTP::Request ();
use HTTP::Response ();


sub init($)
{   my ($self, $args) = @_;
    $self->SUPER::init($args);
    $self->{IMHC_sent} = [];
    $self;
}

#--------------

sub msgsSent()   {shift->{IMHC_sent}}

#--------------

sub sendRequest($$$)
{   my ($self, $req, $cb, $session) = @_;
    $req->header(Accept_Transfer_Encoding => 'chunked, 8bit');
    $req->protocol('HTTP/1.1');

    push @{$self->{IMHC_sent}}, [$req, $cb, $session];
    $self->sendMessage($req, sub {
       # message sent completed
       });
}

sub headerArrived($)
{   my $self = shift;
    HTTP::Response->parse(shift);
}

sub messageArrived($)
{   my ($self, $resp) = @_;
    my $waiting = shift @{$self->{IMHC_sent}};
    unless($waiting)
    {   alert "message arrived, but there was no request";
        return;
    }
    my ($req, $cb, $session) = @$waiting;
    $cb->($self, $req, $resp->code, $resp, $session);
}


1;


lib/IO/Mux/HTTP/Service.pm
# Copyrights 2011 by Mark Overmeer.
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 1.07.
use warnings;
use strict;

package IO::Mux::HTTP::Service;
use vars '$VERSION';
$VERSION = '0.10';

use base 'IO::Mux::HTTP';

use Log::Report 'io-mux-http';

use HTTP::Request  ();
use HTTP::Response ();
use HTTP::Status;
use Socket;


my $conn_id = 'C0000000';

sub init($)
{   my ($self, $args) = @_;
    $args->{name} ||= ++$conn_id;

    $self->SUPER::init($args);
    $self->{IMHS_requests} = [];
    $self->{IMHS_handler}  = $args->{handler}
        or error __x"service {name} is started without handler callback"
             , name => $self->name;
    $self->{IMHS_session}  = {};
    $self->{IMHS_sent}     = [];
    $self;
}

#---------------------

sub client()  {shift->{IMHS_client}}
sub session() {shift->{IMHS_session}}
sub msgsSent(){shift->{IMHS_sent}}

# called when this object gets connected to the mux
sub mux_init($)
{   my ($self, $mux) = @_;
    $self->SUPER::mux_init($mux);

    my $peername         = $self->socket->peername;
    my ($port, $addr)    = unpack_sockaddr_in $peername;
    my $ip               = inet_ntoa $addr;
    my $host; # would be nice to have a async dnslookup here
    my %client           = (port => $port, ip => $ip, host => $host);
    $self->{IMHS_client} = \%client;
}

sub headerArrived($)
{   my $self  = shift;
    HTTP::Request->parse(shift);
}

sub bodyComponentArrived($$)
{   my ($self, $req, $refdata) = @_;

    my $headers = $req->headers;
    my $te = lc($headers->header('Transfer-Encoding') || '8bit');
    return $self->SUPER::bodyComponentArrived($req, $refdata)
        if $te eq '8bit';

    if($te ne 'chunked')
    {   trace "Unsupported transfer encoding $te";
        return $self->errorResponse($req, RC_NOT_IMPLEMENTED);
    }

    my ($starter, $len) = $$refdata =~ m/^((\S+)\r?\n)/ or return;
    if($len !~ m/^[0-9a-fA-F]+$/)
    {   trace "Bad chunk header $len";
        return $self->errorResponse($req, RC_BAD_REQUEST);
    }

    my $need = hex $len;
    my $chunk_length = length($starter) + $need + 2;
    return  # need more data for chunck
        if length($$refdata) < $chunk_length;
 
    if($need!=0)
    {   $req->add_content(substr $$refdata, length($starter), $need, '');
        return;  # get more chunks
    }

    return if $$refdata !~ m/\n\r?\n/;  # need footer
    my ($footer) = $$refdata =~ s/^0+\r?\n(.*?\r?\n)\r?\n//;
    my $header   = $req->headers;
    HTTP::Message->parse($footer)->headers
        ->scan(sub { $header->push_header(@_)} );

    $header->_header('Content-Length' => length ${$req->content_ref});
    $header->remove_header('Transfer-Encoding');
    $req;
}

sub messageArrived($;$)
{   my ($self, $req, $resp) = @_;

    if(my $waiting = shift @{$self->{IMHS_sent}})
    {   # try to continue on track
        my ($resp, $cb, $session) = @$waiting;
        return $cb->($self, $resp, $resp->code, $req, $session);
    }

    $self->shutdown(0)      # shutdown on low-level errors
       if $resp;

    unless($resp)
    {   # Auto-reply to "Expect" requests
        my $headers = $req->headers;
        if(my $expect = $headers->header('Expect'))
        {   $resp = lc $expect ne '100-continue'
              ? $self->errorResponse($req, RC_EXPECTATION_FAILED)
              : $self->errorResponse($req, RC_CONTINUE);
        }
    }

    my $queue = $self->{IMHS_requests};
    push @$queue, [$req, $resp];
    # trace "new queued ".$req->uri.'; ql='.@$queue;

    # handler initiated by first request in queue, then auto-continues
    $self->nextRequest
        if @$queue==1;
}

# This is the most tricky part: each connection may have multiple
# requests queued.  If the handler returns a response object, the
# the response succeeded.  Otherwise, other IO will need to be performed:
# we simply stop.  When the other IO has completed, it will call this
# function again, to resolve the other requests.

sub nextRequest()
{   my $self    = shift;
    my $queue   = $self->{IMHS_requests};
    my $starter = $self->{IMHS_handler};

    #trace "nextRequest: ".join(',', map {$_->[0]->uri} @$queue);
    while(@$queue)
    {   my $first = $queue->[0];
        my ($req, $resp) = @$first;
        if($resp)
        {   info "response already prepared: ".$req->uri;
            $self->sendResponse($resp, sub {} );
        }
        else
        {   info "initiate new session: ".$req->uri;
            $starter->($self, $req, $self->{IMHS_session});
        }
        shift @$queue;
    }
}

#--------------

sub sendResponse($$;$)
{   my ($self, $resp, $user_cb, $session) = @_;
    $resp->protocol('HTTP/1.1');
    push @{$self->{IMHS_sent}}, [$resp, $user_cb, $session];
    $self->sendMessage($resp, sub {
        # message send completed
        });
}


sub makeResponse($$$;$)
{   my ($self, $req, $status, $header, $content) = @_;
    my $resp = HTTP::Response->new($status, status_message($status), $header);
    $resp->request($req);

    $content or return $resp;

       if(ref $content eq 'CODE')   { $resp->content($content) }
    elsif(ref $content eq 'SCALAR') { $resp->content_ref($content) }
    else                            { $resp->content_ref(\$content) }

    $resp;
}


sub errorResponse($$;$)
{   my ($self, $req, $status, $text) = @_;
    my $descr   = defined $text && length $text ? "\n<p>$text</p>" : '';
    my @headers = ('Content-Type' => 'text/html');
    my $message = status_message $status;

    $self->makeResponse($req, $status, \@headers, \<<__CONTENT);
<html><head><title>$status $message</title></head>
<body><h1>$status $message</h1>$descr
</body></html>
__CONTENT
}


sub redirectResponse($$$;$)
{   my ($self, $req, $status, $location, $content) = @_;
    is_redirect $status
        or panic "Status '$status' is not redirect";

    my @headers = (Location => $location);
    if(defined $content && length $content)
    {   my $ct  = $content =~ m/^\s*\</ ? 'text/html' : 'text/plain';
        push @headers, 'Content-Type' => $ct;
    }

    $self->makeResponse($req, $status, \@headers, $content);
}

#---------------------


1;
